[DOC] Improve usability of upgrade docs (#1528)
strimzi/strimzi-kafka-operator: documentation/book/assembly-upgrade.adoc

// master.adoc

[id='assembly-upgrade-{context}']

> = Upgrade

Reviewer: I think the title should make it more clear what upgrades are we talking about ... I think the original title was better.

Author: I was going with the other highest level titles like 'Security'. No strong feelings, so have reverted.

> You can upgrade {ProductName} with no cluster downtime. Each version of {ProductName} supports a range of versions of Apache Kafka. You can upgrade to a higher Kafka version as long as that version is supported by your version of {ProductName}.

Reviewer: Range seems to suggest there will be always be more of them ...

```suggestion
You can upgrade {ProductName} with no cluster downtime. Each version of {ProductName} supports a one or more versions of Apache Kafka. You can upgrade to a higher Kafka version as long as that version is supported by your version of {ProductName}.
```

In some cases, you can also downgrade to a lower supported Kafka version.

> Later versions of {ProductName} provide support for newer versions of Kafka, so you need to upgrade {ProductName} _before_ you can upgrade to a higher supported Kafka version.

Reviewer: This sound strange. These are docs for version X.Y. Not for any older versions. So I do not think we should use things like _Later versions_.

Author: Changed to: '{ProductName} provides support for newer versions of Kafka, but you need to upgrade {ProductName} _before_ you can upgrade to a higher supported Kafka version.'

ifdef::StrimziUpgrades[]

> *Kafka versions*

Reviewer: Please change this into a header, so that we can link to them.

> The Kafka message format version and inter-broker protocol version are subject to change. As a result, the upgrade process involves making both configuration changes to existing Kafka brokers and code changes to client applications (consumers and producers). The following table shows the differences between the Kafka versions:

Reviewer: _code changes to client applications (consumers and producers)_ this suggests that users need to change their application code. I do not think this is true.

Author: changed to: _involves making configuration changes to existing Kafka brokers and client applications (consumers and producers)_

include::snip-kafka-versions.adoc[leveloffset=+1]

endif::StrimziUpgrades[]

> ifndef::StrimziUpgrades[]

Reviewer: This is one of the sections which shouldn't really be here because they belong downstream. Should we start doing something about it already?

Author: We are going to plan in the work as separate issues for RHEL and OCP. One thing that could really help is just showing the oc command in the upstream procedures. Do we really have to duplicate all the commands? We could have an upstream snippet at the start of the doc that states that you can use either oc or kubectl, but for 'clarity' we're just showing oc in the examples?

Reviewer: I see two problems with it:
* I'm not sure the commands are always the same
* If we would need to decide for one, it would probably be `kubectl` and not `oc` which might make your life easier.

(BTW: This comment is for me not a blocker for merging this PR - I just wanted to point out that we will need to address it in the future. But I'm fine with merging this.)

Author: Thanks Jakub. We could have kubectl upstream and oc downstream using a variable. We can give it some thought.

Reviewer 2: I still don't understand why we're sticking with the strategy of documenting both platforms in the same book. No one has given a good reason why we shouldn't just produce two books, one for OCP and one for Kubernetes. Many of the examples could just be dealt with via an attribute as @PaulRMellor suggests. It would make the docs easier to maintain and easier for the user to understand.

Reviewer 3: I am a HUGE fan of not having 1 book for everything! I think it was a request early on, but is causing all kinds of issues.  @jenmalloy can talk to having 2 (or more) books from her experience with EnMasse.

Author: I have created a separate issue to track the conversation on the upstream/downstream split:

https://issues.jboss.org/browse/ENTMQST-919

Please add any thoughts there. We can use this PR to get the upgrade content in and treat the re-architecture separately.

> *Kafka versions*

Reviewer: Please change this into a header, so that we can link to them.

{ProductName} 1.0.0 and 1.1.0 are based on different versions of Apache Kafka:

[options="header"]
|=======================
|{ProductName} version |Kafka version
|1.0.0                 |2.0.0
|1.1.0                 |2.1.1
|=======================

The Kafka message format version and inter-broker protocol version are subject to change. As a result, the upgrade process involves making both configuration changes to existing Kafka brokers and code changes to client applications (consumers and producers). The following table shows the differences between the Kafka versions:

[options="header"]
|=======================
|Kafka version |Interbroker protocol version  |Log message format version | Zookeeper version
|2.0.0         |2.0                           |2.0                        | 3.4.13
|2.1.1         |2.1                           |2.1                        | 3.4.13
|=======================

Although Kafka 2.0.0 and 2.1.1 use the same version of Zookeeper, Red Hat recommends that you update your Zookeeper cluster to use the newest Zookeeper binaries before proceeding with the main {ProductName} 1.1.0 upgrade.
endif::StrimziUpgrades[]

> *Message format version*

Reviewer: Please change this into a header, so that we can link to them.

Author: Changed

When a producer sends a message to a Kafka broker, the message is encoded using a specific format. This is referred to as the message format version. You can configure a Kafka broker to convert messages from newer format versions to a given older format version, before the broker appends the message to the log.

In Kafka, there are two different methods for setting the message format version:

* The `log.message.format.version` property is set on Kafka brokers.

* The `message.format.version` property is set on topics.

The default value of `message.format.version` for a  topic is defined by the `log.message.format.version` that is set on the Kafka broker. You can manually set the `message.format.version` of a topic by modifying its topic configuration.

In the upgrade tasks, we assume that the message format version is defined by the `log.message.format.version`.

> *Procedure outline*

Reviewer: Please change this into a header, so that we can link to them.

Author: Changed

> [id='assembly-upgrade-{context}']

Reviewer 3: This assembly has a lot of content in it - it should be mostly includes to modules.  There are any concepts and reference material here that should all be split out.

Author: Agreed. I will do some work to reduce the content down. It might be a separate issue so I can concentrate on improving the content here.

= {ProductName} and Kafka upgrades

You can upgrade {ProductName} with no cluster downtime. Each version of {ProductName} supports one or more versions of Apache Kafka. You can upgrade to a higher Kafka version as long as that version is supported by your version of {ProductName}.
In some cases, you can also downgrade to a lower supported Kafka version.

{ProductName} provides support for newer versions of Kafka, but you need to upgrade {ProductName} _before_ you can upgrade to a higher supported Kafka version.

> *Procedure outline*

Reviewer 3: ```suggestion
.Procedure outline
```

Reviewer 3: Bold should be used to identify GUI items

> In summary, upgrading {ProductName} is a two-stage process. To upgrade brokers and clients without downtime, you *must* complete the upgrade procedures in the order shown here.

Reviewer 3: ```suggestion
In summary, upgrading {ProductName} is a two-stage process. To upgrade brokers and clients without downtime, you _must_ complete the upgrade procedures in the order shown here.
```

. First, update your Cluster Operator to the latest {ProductName} version.

.. xref:assembly-upgrade-cluster-operator-{context}[]

. Second, upgrade all Kafka brokers and client applications to the latest Kafka version.

> .. xref:assembly-upgrading-kafka-versions-{context}[]

Author: The bullets don't look so bad.

ifdef::StrimziUpgrades[]
== Kafka versions

> The Kafka message format version and inter-broker protocol version are subject to change. As a result, the upgrade process involves making configuration changes to existing Kafka brokers and client applications (consumers and producers). The following table shows the differences between the Kafka versions:

Author: Added

include::snip-kafka-versions.adoc[leveloffset=+1]

endif::StrimziUpgrades[]

ifndef::StrimziUpgrades[]

> == Kafka versions

Reviewer 3: This is the same heading title as the one above. Can we change this one or the one on line 25?

Author: You only see one due to the If statements (which will be going). The content in the assembly will be reduced too.

{ProductName} 1.0.0 and 1.1.0 are based on different versions of Apache Kafka:

[options="header"]
|=======================
|{ProductName} version |Kafka version
|1.0.0                 |2.0.0
|1.1.0                 |2.1.1
|=======================

> The Kafka message format version and inter-broker protocol version are subject to change. As a result, the upgrade process involves making both configuration changes to existing Kafka brokers and code changes to client applications (consumers and producers). The following table shows the differences between the Kafka versions:

Author: Kafka's log message format version and inter-broker protocol version specify the version for the log format appended to messages and the broker protocol used in a cluster. As a result, the upgrade process involves making configuration changes to existing Kafka brokers and code changes to client applications (consumers and producers) to ensure the correct versions are used.

[options="header"]
|=======================
|Kafka version |Interbroker protocol version  |Log message format version | Zookeeper version
|2.0.0         |2.0                           |2.0                        | 3.4.13
|2.1.1         |2.1                           |2.1                        | 3.4.13
|=======================

> Although Kafka 2.0.0 and 2.1.1 use the same version of Zookeeper, Red Hat recommends that you update your Zookeeper cluster to use the newest Zookeeper binaries before proceeding with the main {ProductName} 1.1.0 upgrade.

Reviewer 3: We should not say 'Red Hat recommends' This is one area where we agreed it's okay to use the passive voice to allow docs to be shared up and downstream.  It should say "It is recommended..."  This is in the contribution guidelines.

Author: Updated

endif::StrimziUpgrades[]

> *Message format version*

Reviewer 3: ```suggestion
.Message format version
```

When a producer sends a message to a Kafka broker, the message is encoded using a specific format. This is referred to as the message format version. You can configure a Kafka broker to convert messages from newer format versions to a given older format version before the broker appends the message to the log.

In Kafka, there are two different methods for setting the message format version:

* The `log.message.format.version` property is set on Kafka brokers.

* The `message.format.version` property is set on topics.

The default value of `message.format.version` for a  topic is defined by the `log.message.format.version` that is set on the Kafka broker. You can manually set the `message.format.version` of a topic by modifying its topic configuration.

In the upgrade tasks, we assume that the message format version is defined by the `log.message.format.version`.

ifndef::StrimziUpgrades[]
== Upgrade prerequisites

Before you begin the upgrade process, make sure that:

* {ProductName} is installed. For instructions, see xref:getting-started-{context}[].

> * You have read the link:https://access.redhat.com/documentation/en-us/red_hat_amq/7.2/html-single/amq_streams_1.1.0_on_openshift_container_platform_release_notes/index[{ProductName} 1.1.0 on Red Hat Enterprise Linux Release Notes].

Reviewer 3: This should be an attribute and shouldn't be in the upstream.

Author: The if statement keeps it out of the upstream doc, but we're changing that way of doing things. Watch this space!
